### ISSUE TITLE
How to retrain a fully trained model on a new dataset (for domain adaptation)

### ISSUE THREAD

**Issue by abaheti95** [OpenNMT/OpenNMT-py#768]
I have a fully trained Conversation model and I want to retrain the weights of that model on a small domain data for some epochs. So I can use the -train_from opt for this. However, I am confused about how to make sure the vocabulary used is same for the domain data.
For example,
Let's say the original training data has 25k vocabulary size. And the domain data has 21k vocabulary size. Now I want to preprocess the domain data with the vocabulary of training data. How do I do this in OpenNMT? What opts should I use while preprocessing. I have *.train.pt, *.valid.pt and *.vocab.pt files from the preprocessing the original training data with 25K vocab size. (I had used -share_vocab option in that)

**Comment by vermasrijan** [comment]
Hi,

First, you need to preprocess your domain data with the following code given below:
" python preprocess.py -train_src data/src-train1.txt -train_tgt data/tgt-train1.txt -valid_src data/src-val1.txt -valid_tgt data/tgt-val1.txt -save_data data/demo2 "
Here, 'src-train1.txt' , 'tgt-train1.txt' , 'src-val1.txt' and 'tgt-val1.txt' are your training and validation datasets for 'domain data' (having 21K vocab).

Then, you need to use the '-train_from' command as given below: 
" python train.py -data data/demo2 -train_from demo-model1_acc_16.36_ppl_119.22_e27.pt -save_model demo-model1 "
Here, 'demo-model1_acc_16.36_ppl_119.22_e27.pt' is the pre-saved model (having 25K vocab).
And the new model is being saved to 'model1....'.

Does this help?

**Comment by abaheti95** [comment]
Thanks for the response. It did work but it resumed learning rate from previous epoch. Is there a way to reset the learning rate to the initial or some other setting?

**Comment by vermasrijan** [comment]
You may want to refer to this : http://opennmt.net/OpenNMT/training/retraining/#resuming-a-stopped-training 
What you are asking for is surely implementable using torch (lua language), but I doubt if the same functionality is implementable using pytorch (for python language)

**Comment by jbecke** [comment]
@hpulfc answer almost worked for me. Here's what I did to make it work (not sure if all of these are important...)

In train_single.py main():

```python
if opt.train_from:
        logger.info('Loading checkpoint from %s' % opt.train_from)
        checkpoint = torch.load(opt.train_from,
                                map_location=lambda storage, loc: storage)
        model_opt = opt
        for i in range(0, len(checkpoint['optim'].optimizer.param_groups)):
            checkpoint['optim'].optimizer.param_groups[i]['lr'] = opt.learning_rate
            checkpoint['optim'].optimizer.param_groups[i]['weight_decay'] = opt.learning_rate_decay
```
and 
```python
# Build optimizer.
    optim = build_optim(model, opt, None)
```

Then in onmt/utils/optimizers.py build_optim() function:
```python
 if False:
        optim = checkpoint['optim']
        # We need to save a copy of optim.optimizer.state_dict() for setting
        # the, optimizer state later on in Stage 2 in this method, since
        # the method optim.set_parameters(model.parameters()) will overwrite
        # optim.optimizer, and with ith the values stored in
        # optim.optimizer.state_dict()
        for i in range(0,len(optim.optimizer.param_groups)):
            optim.optimizer.param_groups[i]['lr'] = opt.learning_rate
            optim.optimizer.param_groups[i]['weight_decay'] = opt.learning_rate_decay
        saved_optimizer_state_dict = optim.optimizer.state_dict()
        saved_optimizer_state_dict['param_groups'][0]['lr'] = opt.learning_rate
        saved_optimizer_state_dict['param_groups'][0]['weight_decay'] = opt.learning_rate_decay
        print(saved_optimizer_state_dict)
        print(optim.optimizer.state_dict())
    else:
        optim = Optimizer(
            opt.optim, opt.learning_rate, opt.max_grad_norm,
            lr_decay=opt.learning_rate_decay,
            start_decay_steps=opt.start_decay_steps,
            decay_steps=opt.decay_steps,
            beta1=opt.adam_beta1,
            beta2=opt.adam_beta2,
            adagrad_accum=opt.adagrad_accumulator_init,
            decay_method=opt.decay_method,
            warmup_steps=opt.warmup_steps,
            model_size=opt.rnn_size)

 # Stage 1:
    # Essentially optim.set_parameters (re-)creates and optimizer using
    # model.paramters() as parameters that will be stored in the
    # optim.optimizer.param_groups field of the torch optimizer class.
    # Importantly, this method does not yet load the optimizer state, as
    # essentially it builds a new optimizer with empty optimizer state and
    # parameters from the model.

    optim.set_parameters(model.named_parameters())

    if False:
        # Stage 2: In this stage, which is only performed when loading an
        # optimizer from a checkpoint, we load the saved_optimizer_state_dict
        # into the re-created optimizer, to set the optim.optimizer.state
        # field, which was previously empty. For this, we use the optimizer
        # state saved in the "saved_optimizer_state_dict" variable for
        # this purpose.
        # See also: https://github.com/pytorch/pytorch/issues/2830
        optim.optimizer.load_state_dict(saved_optimizer_state_dict)
        # Convert back the state values to cuda type if applicable
        if use_gpu(opt):
            for state in optim.optimizer.state.values():
                for k, v in state.items():
                    if torch.is_tensor(v):
                        state[k] = v.cuda()

        # We want to make sure that indeed we have a non-empty optimizer state
        # when we loaded an existing model. This should be at least the case
        # for Adam, which saves "exp_avg" and "exp_avg_sq" state
        # (Exponential moving average of gradient and squared gradient values)
        if (optim.method == 'adam') and (len(optim.optimizer.state) < 1):
            raise RuntimeError(
                "Error: loaded Adam optimizer from existing model" +
                " but optimizer state is empty")

    return optim

```

Note that I just changed if statements to if False because I'm lazy...

**Comment by kaihuchen** [comment]
@jbecke Thanks for sharing the code. This is very useful to me.

One thing is that this code overrides the default behavior of continuing training from an existing model (parameters and all) using the -train_from option, which I still need. So I added a 'resume' option to toggle between the two behavior, now all's well.

**Comment by jbecke** [comment]
@kaihuchen glad it was useful. I’m not sure what you mean in your second paragraph. The code should still load the old model parameters (i.e. structure, weights and biases) but create a new optimizer with the given params. Could uou clarify what your issue is?

I will make a PR if the maintainers (@srush (?)) agree that this should be the default behavior, i.e. use previous flags but override ones that the user explicitly specifies when they run -train_from?

**Comment by kaihuchen** [comment]
@jbecke 

Hmmm. What I meant above was that with your code change I thought I am losing the trained weights, since when I restarted the session using the train_from option to load a fully trained model (with good accuracy and low perplexity),  I observe the accuracy goes way down to near zero and perplexity shot way up as if the training is starting from scratch. 

Hope this clarifies.

**Comment by fuzihaofzh** [comment]
@vermasrijan Thanks for your answer. But how to deal with new words in the new dataset?

**Comment by cemicel** [comment]
is there going to be an official fix for learning rate (when using train_form option) ?

**Comment by vince62s** [comment]
it is already possible, not sure to understand your point.
